### PR TITLE
viper: adorn invariants with permissions

### PR DESCRIPTION
**Pull request by ggreif**
EXPERIMENTAL, but hey, it's a start!

TODO:
- [ ] `pp_perm`
- [ ] remove duplicates (optimise conjunction expressions)
- [x] harvest `perm`s from statement lists (`write` and _read_)
  - [x] apply to methods
  - [x] apply to `__init__`
- [ ] deduplicate perms and eliminate the trailing `true`.

### CODE REVIEW

File: src/viper/syntax.ml [hunk 1]
```diff
@@ -49,7 +49,7 @@ and perm = (perm', info) Source.annotated_phrase
 
 and perm' =
   | WildcardP
-  | FullP
+  | FullP of fldacc
```
Comment (ggreif) [inline src/viper/syntax.ml:52]: What is `FullP`? :-)

Comment (aterga) [reply src/viper/syntax.ml:52]: So, none of the named values need to carry any additional parameters, whereas we do need to add an arbitrary permission node that would carry two parameters, e.g., the `p` and `q` in `acc(x.f, p/q)`.

**Comment by aterga** [comment]
I created another PR that implement this PR's functionality (without permission inference) https://github.com/dfinity/motoko/pull/3496

**Comment by aterga** [comment]
This PR might be used later for extracting fine-grained permissions from the code.

**Pull request closed without merging**